modules/installation-special-config-kmod.adoc

```diff
@@ -0,0 +1,306 @@
+   Loaded: loaded (/etc/systemd/system/kmods-via-containers@.service;
+          enabled; vendor preset: disabled)
+   Active: active (exited) since Sun 2020-01-12 23:49:49 EST; 5s ago...
+----
+
+. Confirm that the kernel modules loaded. For example:
++
+----
+$ lsmod | grep simple_
+simple_procfs_kmod     16384  0
+simple_kmod            16384  0
+$ dmesg | grep 'Hello world'
+[ 6420.761332] Hello world from simple_kmod.
+$ sudo cat /proc/simple-procfs-kmod
+simple-procfs-kmod number = 0
+$ spkut 44
+KVC: wrapper simple-kmod for 4.18.0-147.3.1.el8_1.x86_64
+Running userspace wrapper using the kernel module container...
++ podman run -i --rm --privileged
+   simple-kmod-dd1a7d4:4.18.0-147.3.1.el8_1.x86_64 spkut 44
+simple-procfs-kmod number = 0
+simple-procfs-kmod number = 44
+----
+
+Going forward, when the system boots this service will check if a new
+kernel is running. If there is a new kernel, the service builds a new
+version of the kernel module and then loads it. If the module is already
+built, it will just load it.
+
+== Provision the kernel module via Ignition
+Depending on whether or not you need to have the kernel module in place
+when {product-title} cluster first boots, you can set up the
+kernel modules to be deployed in one of two ways:
+
+* **Provision kernel modules at cluster install time (day-1)**: You can add
+the kernel module software to Ignition config files during installation by
+including the required content through the `openshift-install` command.
+
+* **Provision kernel modules via MCO (day-2)**: If you can wait until the
+cluster is up and running to add your kernel module, you can the kernel
+module software via the Machine Config Operator (MCO).
+
+In either case, each node needs access to new kernel packages and related
+software packages as they are available. There are a few ways to provide this information:
+
+* Provide RHEL entitlements to each node.
+* Get RHEL entitlements from an existing RHEL host, from the `/etc/pki/entitlement` directory
+and copy them to the same location in `$FAKEROOT` when you build your Ignition config.
+* Inside the Dockerfile, add pointers to a yum repository containing the kernel and other packages.
+This must include new kernel packages as they are available.
+
+=== Provision kernel modules at the time of cluster installation
+
+First create a base Ignition config that you would like to use. It will
+contain the ssh public key to add to the authorized_keys file for
+the core user and also a systemd unit file. The systemd unit file
+require-simple-kmod.service that requires kmods-via-containers@simple-kmod.service.
+
+The systemd unit is a workaround for an
+link:https://github.com/coreos/ignition/issues/586[upstream bug]
+and makes sure that the kmods-via-containers@simple-kmod.service gets started
+on boot:
+
+. Create an Ignition config file that creates a systemd unit file:
++
+----
+$ cat <<EOF > ./baseconfig.ign
+{
+  "ignition": { "version": "3.0.0" },
```

Comment:
Since this example is going to be used with RHCOS, the `version` string should be `2.2.0`

Reply:
Done